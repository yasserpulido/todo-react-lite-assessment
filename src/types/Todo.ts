export interface Todo {
  id: string;
  value: string;
  isDone: boolean;
}
