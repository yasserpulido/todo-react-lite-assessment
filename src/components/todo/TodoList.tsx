/** @jsxImportSource @emotion/react */
import React, { useState } from "react";
import { BaseSyntheticEvent } from "../../types/Events";
import { Todo } from "../../types/Todo";
import EditTodo from "./EditTodo";
import { css } from "@emotion/react";

interface Props {
  todos: Todo[];
  doneToggle: (a: string, b: boolean) => void;
  editTodo: (a: string, b: string) => void;
}

const listStyle = css`
  list-style: none;
`;

const labelCheck = css`
  text-decoration-line: line-through;
`;

const TodoList: React.FC<Props> = (props) => {
  const [showEdit, setShowEdit] = useState({ id: "", isEdit: false });

  const doneHandler = (event: BaseSyntheticEvent) => {
    const { id, checked } = event.target;
    props.doneToggle(id, checked);
  };

  const editTodoHandler = (id: string, value: string) => {
    props.editTodo(id, value);
    setShowEdit({ id: "", isEdit: false });
  };

  const showEditHandler = (todo: Todo) => {
    setShowEdit((prevState) => {
      return { id: todo.id, isEdit: !prevState.isEdit };
    });
  };

  return (
    <ul css={listStyle}>
      {props.todos.map((todo) => (
        <li key={todo.id}>
          <input
            type="checkbox"
            id={todo.id}
            checked={todo.isDone}
            onChange={doneHandler}
          />
          <label
            css={todo.isDone ? labelCheck : ""}
            onClick={() => showEditHandler(todo)}
          >
            {todo.value}
          </label>
          {showEdit.isEdit && showEdit.id === todo.id && (
            <EditTodo todo={todo} editTodo={editTodoHandler} />
          )}
        </li>
      ))}
    </ul>
  );
};

export default TodoList;
