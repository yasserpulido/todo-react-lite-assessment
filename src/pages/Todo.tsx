/** @jsxImportSource @emotion/react */
import React, { useState } from "react";
import NewTodo from "../components/todo/NewTodo";
import TodoList from "../components/todo/TodoList";
import { Todo as TodoType } from "../types/Todo";
import { css } from "@emotion/react";

const DUMMY_DATA: TodoType[] = [
  {
    id: "t01",
    value: "Aprender TypeScript",
    isDone: false,
  },
  {
    id: "t02",
    value: "Hacer Tarea",
    isDone: true,
  },
  {
    id: "t03",
    value: "Entregar Tarea",
    isDone: false,
  },
];

const todoCard = css`
  margin: 4rem auto;
  padding: 2rem 3rem 3rem;
  max-width: 500px;
  background-color: #fbc531;
  color: #000;
  box-shadow: 5px 5px 0px 0px rgb(0, 0, 0, 1);
  border: 2px solid #000;
  border-radius: 4px;
`;

const todoTitle = css`
  text-align: center;
`;

const showDoneButton = css({
  backgroundColor: "#f5f6fa",
  border: "1px solid #95a5a6",
  cursor: "pointer",
  padding: "6px 16px",
  borderRadius: "4px",
  marginBottom: "8px",
  textAlign: "right",
  "&:hover": {
    backgroundColor: "#dcdde1",
  },
});

const formWrap = css({
  textAlign: "right",
});

const Todo: React.FC = () => {
  const [todos, setTodos] = useState(DUMMY_DATA);
  const [showDone, setShowDone] = useState(false);

  const newTodoHandler = (todo: string): void => {
    const idGenerated = Math.floor(Math.random() * 100).toString();

    setTodos((prevState: TodoType[]) => {
      return [...prevState, { id: idGenerated, value: todo, isDone: false }];
    });
  };

  const doneToggleHandler = (id: string, checked: boolean): void => {
    const n = todos;

    n.forEach((value) => {
      if (value.id === id) {
        value.isDone = checked;
      }
    });

    setTodos(() => {
      return [...n];
    });
  };

  const editTodoHandler = (id: string, value: string) => {
    const updatedTodo = todos.map((todo) => {
      if (todo.id === id) {
        return { ...todo, value };
      }
      return todo;
    });

    setTodos(updatedTodo);
  };

  const showDoneHandler = () => {
    setShowDone(!showDone);
  };

  const todosFilter = showDone
    ? todos
    : todos.filter((todo) => todo.isDone === showDone);

  return (
    <div css={todoCard}>
      <h1 css={todoTitle}>LISTA DE TAREAS</h1>
      <div css={formWrap}>
        <button css={showDoneButton} type="button" onClick={showDoneHandler}>
          {!showDone ? "Mostrar Completadas" : "Quitar Completadas"}
        </button>
        <NewTodo newTodo={newTodoHandler} />
      </div>
      <TodoList
        todos={todosFilter}
        doneToggle={doneToggleHandler}
        editTodo={editTodoHandler}
      />
    </div>
  );
};

export default Todo;
