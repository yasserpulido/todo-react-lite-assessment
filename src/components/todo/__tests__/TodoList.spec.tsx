import { render, screen, within } from "@testing-library/react";
import { Todo } from "../../../types/Todo";
import TodoList from "../TodoList";

describe("TodoList", () => {
  const todo: Todo[] = [
    {
      id: "t01",
      text: "Hacer una prueba",
      done: false,
    },
    {
      id: "t02",
      text: "Hacer dos prueba",
      done: false,
    },
  ];

  test("Should TodoList to match Snapshot", () => {
    // Act
    const { container } = render(
      <TodoList todos={todo} doneToggle={() => {}} updateTodo={() => {}} />
    );

    // Assert
    expect(container).toMatchSnapshot();
  });

  test("Should TodoList render", () => {
    // Act
    render(
      <TodoList todos={todo} doneToggle={() => {}} updateTodo={() => {}} />
    );
    const list = screen.getByRole("list");
    const { getAllByRole } = within(list)
    const items = getAllByRole("listitem")

    // Assert
    expect(list).toBeInTheDocument();
    expect(items.length).toBe(2)
  });
});
