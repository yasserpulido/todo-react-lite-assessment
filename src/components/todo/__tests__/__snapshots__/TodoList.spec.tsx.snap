// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Should TodoList to match Snapshot 1`] = `
<div>
  <ul
    class="css-r3gwee-TodoList"
  >
    <li>
      <input
        id="t01"
        type="checkbox"
      />
      <label
        class="css-0"
      >
        Hacer una prueba
      </label>
    </li>
    <li>
      <input
        id="t02"
        type="checkbox"
      />
      <label
        class="css-0"
      >
        Hacer dos prueba
      </label>
    </li>
  </ul>
</div>
`;
